partiql-ast/README.adoc

```diff
@@ -0,0 +1,356 @@
+
+    public val children: List<AstNode>
+
+    public fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R
+}
+----
+
+=== Example
+
+.Example Definition
+[source,ion]
+----
+expr::[
+ // ...
+ binary::{
+   op: [
+     PLUS, MINUS, TIMES, DIVIDE, MODULO, CONCAT,
+     AND, OR,
+     EQ, NE, GT, GTE, LT, LTE,
+   ],
+   lhs: expr,
+   rhs: expr,
+ },
+ // ...
+]
+----
+
+.Generated Interface
+[source,kotlin]
+----
+// Note: `Expr:AstNode` is a sealed interface of all expr variants
+
+public interface Binary : Expr {
+    public val op: Op
+    public val lhs: Expr
+    public val rhs: Expr
+
+    public fun copy(
+      op: Op = this.op,
+      lhs: Expr = this.lhs,
+      rhs: Expr = this.rhs,
+    ): Binary
+
+    public enum class Op {
+      PLUS,
+      MINUS,
+      TIMES,
+      DIVIDE,
+      MODULO,
+      CONCAT,
+      AND,
+      OR,
+      EQ,
+      NE,
+      GT,
+      GTE,
+      LT,
+      LTE,
+    }
+}
+----
+
+== Factory, DSL, and Builders
+
+The PartiQL AST library provides several creational patterns in `org.partiql.ast.builder` such as an abstract base factory, Kotlin DSL, and Java fluent-builders.
+These patterns enable customers to extend the AST to fit their needs, while providing a base which can be decorated appropriately.
+
+=== Factory Usage
+
+The factory is how you instantiate a node. The default factory can be called directly like,
+
+[source,kotlin]
+----
+Ast.exprLit(int32Value(1))  // expr.lit
+----
+
+==== Custom Nodes
+
+Additionally, you can extend the abstract base factory and use it in builders as well as the DSL. This gives you full
+control over how your nodes are instantiated. If you are ambitious, you can implement your own versions of AST node interfaces and implement a base factory. This
+will allow you to create custom behaviors. For example, generated equals functions do not consider semantics. Perhaps
+we want to improve how we compare nodes? Here's an example that considers the case-sensitivity of identifiers.
+
+.Custom Node and Factory Example
+[source,kotlin]
+----
+public abstract class MyFactory : AstBaseFactory() {
+
+    override fun identifierSymbol(symbol: String, caseSensitivity: Identifier.CaseSensitivity): Identifier.Symbol {
+        return ComparableIdentifier(_id(), symbol, caseSensitivity)
+    }
+}
+
+class ComparableIdentifier(
+    override val _id: String,
+    override val symbol: String,
+    override val caseSensitivity: Identifier.CaseSensitivity,
+) : Identifier.Symbol {
+
+    // override copy, children
+
+    override fun equals(other: Any?): Boolean {
+        if (other == null || other !is Identifier.Symbol) return false // different type
+        if (other === this) return true // same object
+        return when (caseSensitivity) {
+            Identifier.CaseSensitivity.SENSITIVE -> this.symbol == other.symbol
+            Identifier.CaseSensitivity.INSENSITIVE -> this.symbol.lowercase() == other.symbol.lowercase()
+        }
+    }
+}
+----
```

Comment:
I used this as an example because we had discussed this exact limitation last fall. This example shows how the current design is able to address situations in which generated sources do not handle edge cases.

I believe it is useful to provide customers with the facility to produce their own concrete implementations of AST nodes (if they want, hence a default). Here we are indeed eating our own dogfood while overcoming basic limitations that are trivial when hand-writing ASTs. The default factory is _our_ default factory, but not _the_ default factory.

> Use of [the abstract factory] pattern enables interchangeable concrete implementations without changing the code that uses them, even at run. [wikipedia](https://en.wikipedia.org/wiki/Abstract_factory_pattern)


Is your primary concern that this is a bad example?

Reply:
My foremost concern is understanding what ready-to-roll uses we have for the significant extra complexity that Abstract Factory pattern introduces.  Not nice-to-haves that a future customer might find useful, but 2-3 usecases (that is, more than 1!) that we intend to roll out ourselves -- otherwise we won't prove the dog food is edible.

The prototypical usecase for Abstract Factory is making it possible to develop a GUI application, such as a game, while abstracting away from a specific windowing toolkit, so that the application can run with with the look and feel of any toolkit for which someone would provide a concrete factory. 

PartiQL AST (the interfaces in partiql-ast/org.partiql.ast.Types) is the analog of a "windowing toolkit abstraction", while PartiQL parser and PartiQL compiler are analogues of a "GUI game".  What would be useful analogues of the concrete windowing toolkits that we are going to roll out as alternatives to the one concrete AST in partiql-ast/org.partiql.ast.**impl**.Types ?  

I can vaguely imagine something on a scale of an AST variant that contains something like instrumentation to count uses of nodes during compilation or optimization transformations -- but I don't have a defensible case for this and don't recall us planning for anything of a kind.  The case-sensitivity example is not convincing because it's something we should just provide (or not!) in the default AST implementation -- case sensitivity ought to be part of the PartiQL AST contract (or not! -- taken care of outside the AST, as currently). 

To put it differently, what do we give up if org.partiql.ast.Types and org.partiql.ast.**impl**.Types are one and the same thing (just concrete classes, no interfaces)?

This probably requires a discussion more sprawling than can fit in PR comments...

Reply:
I'm aware of the pattern background, and I see your concern now. The motivation for this was to have the flexibility of hand-written AST nodes, but the convenience of generated ones. This pattern enables us to get everything generated, but hand-write when we desire. The default ~GUI~ AST toolkit is what we provide, but customers can provide their own ~look and feel~ nodes when desired/required. 

As you'll see in the PIG AST and the Physical Plan. Various mechanisms have been added for customer extension such as the metadata `<string, object>` map as well as the `impl` string on the `partiql_physical` domain (which is a point of customer extension, but is also related to the "physical"-ness of the domain). 

My effort here was to provide a more powerful mechanism for both ourselves and customers to extend AST nodes. That is, they can use the generated identifiers to store metadata in a richer-typed structure than the object map (see SourceLocations) for an example. And they can handwrite implementations of our AST node interfaces.

My opinion is that, if I could go back in time, the AST would be internal and hand-written. It is the driving cause of major-version churn due to its constant evolving and surface area, as we as its inflexibility due to the generator limitations. I don't this PR solves that issue, but it gives us some better defaults and flexibility. If the AST were internal, then the separation of interfaces and indirection of instantiation would be superfluous. But I found this pattern to be a reasonable compromise given our situation.